Code/RDGeneral/tags-threadsafe.h

```diff
@@ -0,0 +1,121 @@
+/* 
+*
+*  Copyright (c) 2016, Novartis Institutes for BioMedical Research Inc.
+*  All rights reserved.
+* 
+* Redistribution and use in source and binary forms, with or without
+* modification, are permitted provided that the following conditions are
+* met: 
+*
+*     * Redistributions of source code must retain the above copyright 
+*       notice, this list of conditions and the following disclaimer.
+*     * Redistributions in binary form must reproduce the above
+*       copyright notice, this list of conditions and the following 
+*       disclaimer in the documentation and/or other materials provided 
+*       with the distribution.
+*     * Neither the name of Novartis Institutes for BioMedical Research Inc. 
+*       nor the names of its contributors may be used to endorse or promote 
+*       products derived from this software without specific prior written permission.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+* A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+* OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+* SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+* LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+* DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+* THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+#ifndef RDKIT_TAGS_THREADSAFE_H
+#define RDKIT_TAGS_THREADSAFE_H
+
+#include "BoostStartInclude.h"
+#include <boost/unordered_map.hpp>
+#include <boost/thread/mutex.hpp>
+#include <boost/thread/tss.hpp>
+#include "BoostEndInclude.h"
+
+#include "types.h"
+
+namespace RDKit
+{
+
+class RDTags
+{
+  RDTags(const RDTags&);
+  RDTags& operator=(const RDTags&);
+public:
+  RDTags() {
+    // Load the known keys
```

Comment:
Sorry, munged a few rambling things into that comment. 

1) Don't add a lock acquisition in the constructor here, it's entirely un-necessary. Even Java that is heavily multi-threaded with how a JVM wide lock during a System.loadLibrary call. 

2) I generally don't like adding any link-time overhead as it is un-avoidable for users and impossible to optimize away for them. Another benefit of the Meyer's singleton pattern is users only have to incur the link-time overhead if they actually use the global shared state. "python -c 'from rdkit import Chem'" is already ~0.23 seconds which I consider kind of high. Run into too many shell scripts running across cluster to invoke a single RDKit function. 

3) Unfortunately, we have to store the interned strings below common_properties::MAX in the unordered_map because tagmap.get(const std::string &name) needs to be able to look up the integer. If we really needed to optimize for #2 later we could reorder the common_properties values to be binary searchable. But that's an optimization for later I feel, so I retract that comment. :-)